### PR TITLE
Add support for ELF files with no. of sections > SHN_LORESERVE (issue #67).

### PR DESCRIPTION
**Pull request by paulhuggett**
This is a fix for issue #67 (“Add proposal 74 support to repo2obj”). The problem was seen when performing multiple builds of LLVM whilst investigating issue #73 (“Adding a new section broke xfixups”).

The fix itself is reasonably straightforward, but testing remains a problem for repo2obj because we don’t yet have the tooling to easily allow a specific set of structures to be created in the database from a textual description (that’s [pstore issue #15](https://github.com/SNSystems/pstore/issues/15)). I do have a python utility which writes an LL file which is then compiled [here](https://gist.github.com/paulhuggett/015c791c7b4778aa67d38c0b6bb2cfed), but this is very slow (> 5 minutes for a debug build of LLVM). Any opinions?

### CODE REVIEW

File: llvm/tools/repo2obj/repo2obj.cpp [hunk 1]
```diff
@@ -539,23 +594,34 @@ int main(int argc, char *argv[]) {
...
   }
 
-  uint64_t SectionHeadersOffset = State.writeSectionHeaders(OS);
-  Header.e_shoff = SectionHeadersOffset;
-  Header.e_shnum = State.SectionHeaders.size();
+  State.setEShNum(Header);
+  Header.e_shoff = State.writeSectionHeaders(OS);
+  static_assert(SectionIndices::StringTab < llvm::ELF::SHN_LORESERVE,
+                "String table index should be less than LORESERVE");
   Header.e_shstrndx = SectionIndices::StringTab;
```
Comment (MaggieYingYi) [inline llvm/tools/repo2obj/repo2obj.cpp]: In the document, it is mentioned `ELF Header: e_shstrndx`:

> If` the section name string table index is greater than SHN_LORESERVE (0xff00), this member has the value SHN_XINDEX (0xffff), and the actual index of the section name string table is in the member sh_link of the section header at index 0. (Both entries will be zero if there is no section name string table.) 

Should the `Ehdr.e_shstrndx` look like below?

```
if (the section name string table index  >= SHN_LORESERVE)
    Ehdr.e_shstrndx = SHN_XINDEX;
else
    Ehdr.e_shstrndx = the section name string table index;
```

File: llvm/tools/repo2obj/R2OELFOutputSection.h [hunk 2]
```diff
@@ -166,8 +166,8 @@ template <typename ELFT> class OutputSection {
   llvm::Optional<std::vector<uint8_t>> const &Prefix_;
   GroupInfo<ELFT> *Group_ = nullptr;
 
-  static constexpr auto UnknownIndex = std::numeric_limits<unsigned>::max();
-  unsigned Index_ = UnknownIndex; // The section header table index
+  static constexpr auto UnknownIndex = std::numeric_limits<size_t>::max();
+  size_t Index_ = UnknownIndex; // The section header table index
```
Comment (MaggieYingYi) [inline llvm/tools/repo2obj/R2OELFOutputSection.h]: Why is `size_t` type not `uint32_t` type?

File: llvm/tools/repo2obj/repo2obj.cpp [hunk 3]
```diff
@@ -69,16 +69,14 @@ class SpecialNames {
 public:
   void initialize(const pstore::database &Db, GeneratedNames &Names);
 
-  pstore::typed_address<pstore::indirect_string> CtorName =
-      pstore::typed_address<pstore::indirect_string>::null();
-  pstore::typed_address<pstore::indirect_string> DtorName =
-      pstore::typed_address<pstore::indirect_string>::null();
+  using istring = pstore::typed_address<pstore::indirect_string>;
+  istring CtorName = istring::null();
+  istring DtorName = istring::null();
```
Comment (MaggieYingYi) [inline llvm/tools/repo2obj/repo2obj.cpp]: Is the code refactoring only here? If yes, could you please put into the different commit.

File: llvm/tools/repo2obj/repo2obj.cpp [hunk 4]
```diff
@@ -99,19 +97,14 @@ void SpecialNames::initialize(const pstore::database &Db, GeneratedNames &Names)
...
-  return (Pos != NameIndex.end(Db))
-             ? pstore::typed_address<pstore::indirect_string>(Pos.get_address())
-             : pstore::typed_address<pstore::indirect_string>::null();
+auto SpecialNames::findString(pstore::database const &Db,
+                              pstore::index::name_index const &NameIndex,
+                              pstore::indirect_string const &Str) -> istring {
+  auto const Pos = NameIndex.find(Db, Str);
+  return (Pos != NameIndex.end(Db)) ? istring{Pos.get_address()}
+                                    : istring::null();
 }
```
Comment (MaggieYingYi) [inline llvm/tools/repo2obj/repo2obj.cpp]: Again, Is the code refactoring only here? If yes, could you please put into the different commit.

**Comment by paulhuggett** [comment]
> We must use LLVM IR as input. I think we still need to add a LLVM test even your python utility takes a long time.

The use of LLVM IR is a real problem for repo2obj tests. We‘re feeding a language designed for compiler optimization into the front-end of the tools and expecting a very specific layout in the database for repo2obj to consume. Not only is this slow, but the tests will be fragile because the output may easily change in response to unrelated changes almost anywhere inside the compile. As if that isn’t bad enough, there are aspects of the repo layout that can’t be expressed in LLVM IR (e.g. multiple sections in a fragment, internal vs external fixups).

This is really a terrible idea!

For example, in this test I need to create a ELF file with a large number of sections. The only way that I can do this is by creating a function which calls a large number of linkonce functions. Without the former, the latter are all optimized away.

If we’re going to be adding many more tests for repo2obj then we need to urgently find a proper solution.

Having said that, if you think that it is an important test for me to implement then I will do so.

File: llvm/tools/repo2obj/repo2obj.cpp (hunk 1 above)
Comment (paulhuggett) [reply llvm/tools/repo2obj/repo2obj.cpp]: > Should the Ehdr.e_shstrndx look like below?

It’s not necessary because `SectionIndices::StringTab` is a small constant. Line 623 holds a static assertion that the value is < `SHN_LORESERVE`.

File: llvm/tools/repo2obj/R2OELFOutputSection.h (hunk 2 above)
Comment (paulhuggett) [reply llvm/tools/repo2obj/R2OELFOutputSection.h]: Because the value holds the `size_t` result of calling `State.SectionHeaders.size()` (see repo2obj.cpp line 590). At some point this will be truncated to 32 bits for the ELF file, but I want to keep the internal state consistent.

File: llvm/tools/repo2obj/repo2obj.cpp (hunk 3 above)
Comment (paulhuggett) [reply llvm/tools/repo2obj/repo2obj.cpp]: It's adding a typedef (”refactoring” seems like a rather grand term). It was intended to make the code easier to read but if it offends you, I’ll remove it.

File: llvm/tools/repo2obj/repo2obj.cpp (hunk 4 above)
Comment (paulhuggett) [reply llvm/tools/repo2obj/repo2obj.cpp]: As above.

**Comment by paulhuggett** [comment]
(I should note that the LLVM IR source file produced by the [script](https://gist.github.com/paulhuggett/015c791c7b4778aa67d38c0b6bb2cfed) runs to 22Mb of text!)

**Comment by paulhuggett** [comment]
Test added. However I have serious reservations about it as described above.

File: llvm/tools/repo2obj/repo2obj.cpp (hunk 1 above)
Comment (MaggieYingYi) [reply llvm/tools/repo2obj/repo2obj.cpp]: Thanks, I see.

File: llvm/tools/repo2obj/repo2obj.cpp (hunk 4 above)
Comment (MaggieYingYi) [reply llvm/tools/repo2obj/repo2obj.cpp]: Thanks for removing this from this commit. Please feel free to submit your change as a separate commit.

**Comment by MaggieYingYi** [comment]
> 
> 
> Test added. However I have serious reservations about it as described above.

I have run this test. It did take a long long ... time to pass the test. I agree with you that we shouldn't add this test.

**Comment by paulhuggett** [comment]
It’s no slower than the code I mentioned in the [original post](https://github.com/SNSystems/llvm-project-prepo/pull/74#issue-378857296). What changed your mind?
